typing/jkind.ml

```diff
@@ -2278,32 +2283,207 @@ let for_boxed_variant cstrs =
+              let existentials =
+                !constructor_unbound_type_vars_excluding_row_variables cstr
+                |> Btype.TypeSet.to_seq
+                |> Seq.map Types.Transient_expr.type_expr
+                |> List.of_seq
+              in
+              let tof_kinds =
+                List.map
+                  (fun ty ->
+                    match get_desc ty with
+                    | Tof_kind _ ->
+                      (* We shouldn't be able to hit this, but it's harmless and
+                         defensive to just keep these types the same. *)
+                      ty
+                    | Tvar { jkind; _ } | Tunivar { jkind; _ } ->
+                      Btype.newgenty (Tof_kind jkind)
+                    | _ ->
+                      Misc.fatal_error
+                        "constructor_unbound_type_vars must return Tvar or \
+                         Tunivar")
+                  existentials
+              in
+              match Types.get_desc res with
+              | Tconstr (_, args, _) ->
+                List.fold_left2
+                  (fun (`Args ret_args, `Params params, seen) arg param ->
+                    if Btype.TypeSet.mem arg seen
+                    then
+                      (* We've already seen this type parameter, so don't add it again.
+                         See Step 2a from Note [With-bounds for GADTs] *)
+                      `Args ret_args, `Params params, seen
+                    else
+                      match Types.get_desc arg, Types.get_desc param with
+                      | Tvar _, Tvar _ ->
+                        (* Only add types which are direct variables. Note that types
+                           which aren't variables might themselves /contain/ variables; if
+                           those variables don't show up on another parameter, they're
+                           treated as existentials. See step 2b from Note [With-bounds for
+                           GADTs] *)
+                        ( `Args (arg :: ret_args),
+                          `Params (param :: params),
+                          Btype.TypeSet.add arg seen )
+                      | _ -> `Args ret_args, `Params params, seen)
+                  ( `Args (existentials @ ret_args),
+                    `Params (tof_kinds @ params),
+                    seen )
+                  args decl_params
+              | _ -> Misc.fatal_error "cd_res must be Tconstr")
+          in
+          { Variant_parts.cstr_arg_tys;
+            cstr_arg_modalities;
+            params;
+            ret_args;
+            seen
+          })
+        Variant_parts.empty cstrs
+    in
+    let cstr_arg_tys =
+      if Misc.Stdlib.List.is_empty params
+      then cstr_arg_tys
+      else
+        match
+          (* Step 2 from Note [With-bounds for GADTs] *)
+          type_apply ret_args
+            (Btype.newgenty
+               (Ttuple (List.map (fun ty -> None, ty) cstr_arg_tys)))
```

Comment:
I don't think it's necessary to build a tuple here. Why not just map `type_apply`? All of the free variables in `cstr_arg_tys` should be listed in `ret_args` and so there is no worrying about sharing. (except maybe row variables?)